bot/exts/fun/madlibs.py

```diff
@@ -0,0 +1,124 @@
+import json
+from asyncio import TimeoutError
+from pathlib import Path
+from random import choice
+from typing import TypedDict
+
+import discord
+from discord.ext import commands
+
+from bot.bot import Bot
+from bot.constants import Colours, NEGATIVE_REPLIES
+
+TIMEOUT = 60.0
+
+
+class MadlibsTemplate(TypedDict):
+    """Structure of a template in the madlibs JSON file."""
+
+    title: str
+    blanks: list[str]
+    value: list[str]
+
+
+class Madlibs(commands.Cog):
+    """Cog for the Madlibs game."""
+
+    def __init__(self, bot: Bot):
+        self.bot = bot
+        self.templates = self._load_templates()
+        self.edited_content = {}
+
+    @staticmethod
+    def _load_templates() -> list[MadlibsTemplate]:
+        madlibs_stories = Path("bot/resources/fun/madlibs_templates.json")
+
+        with open(madlibs_stories) as file:
+            return json.load(file)
+
+    @staticmethod
+    def madlibs_embed(part_of_speech: str, number_of_inputs: int) -> discord.Embed:
+        """Method to have the bot send an embed with the game information."""
+        madlibs_embed = discord.Embed(title="Madlibs", color=Colours.python_blue)
+
+        madlibs_embed.add_field(
+            name="Enter a word that fits the given part of speech!",
+            value=f"Part of speech: {part_of_speech}"
+        )
+
+        madlibs_embed.set_footer(text=f"Inputs remaining: {number_of_inputs}")
+
+        return madlibs_embed
+
+    @commands.Cog.listener()
+    async def on_message_edit(self, _: discord.Message, after: discord.Message) -> None:
+        """A listener that checks for edits to messages from the user."""
+        self.edited_content[after.id] = after.content
```

Comment:
We need some form of if-statement so that not all messages get added here.

I have an idea, in discord.py there is a way to say how many commands can get used at once. If we use this we can limit the amount of commands to one at a time, then completely reset this dictionary each time the command ends.

Maybe with a before and after hook we can setup so that some form of attribute gets set that says that a command is ongoing? That way it'll run even if the command errors out.

Reply:
Can't I just use the `author_check` you added to my main program? @Bluenix2

Reply:
This listener is not aware of that author check.

Reply:
Well right now it isn't of course, but could I add it here? Would it work? @Bluenix2

Reply:
Sadly no, because there's no `ctx`.

Reply:
Ahhh, ok. Well could I add it as a parameter to the function? Would that work? @Bluenix2

Reply:
No that wouldn't work either because discord.py creates the `ctx` for each command. So we need to somehow tell the listener each command running - kind of like how/what I suggested!

Reply:
Ok, I see. I'll think about what you suggested then and if I have any questions, I will ask. Thanks!